File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Areas/Backoffice/Models/CategoriesService.vb
Public Class CategoriesService
    Public Shared Function GetCategories() As List(Of Category)
        Dim context = HttpContext.Current

        If context Is Nothing Then
            Throw New InvalidOperationException("A valid HttpContext is required")
        End If

        Dim res = TryCast(context.Cache("categories.all"), List(Of Category))
        If res Is Nothing Then
            res = ObjectContextModule.CurrentContext.CategorySet.ToList
            context.Cache.Insert("categories.all", res, Nothing,
                                 DateTime.Now.AddHours(1), Cache.NoSlidingExpiration)
        End If

        Return res
    End Function
End Class

File: aspnet-in-practice/08-09/CS/CoolMVBlog/Areas/Backoffice/Models/CategoriesService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using CoolMvcBlog.Models;
using System.Web.Caching;
using CoolMvcBlog.Utils;

namespace CoolMvcBlog.Areas.Backoffice.Models
{
    public class CategoriesService
    {
        public static List<Category> GetCategories()
        {
            HttpContext context = HttpContext.Current;
            if (context == null)
                throw new InvalidOperationException("A valid HttpContext is needed");

            var res = context.Cache["categories.all"] as List<Category>;
            if (res == null)
            {
                context.Cache.Insert("categories.all", 
                    res = ObjectContextModule.CurrentContext.CategorySet.ToList(),
                    null, DateTime.Now.AddHours(1), Cache.NoSlidingExpiration);
            }

            return res;
        }
    }
}